### PR TITLE
Add support for setting a tab's color at runtime w/ context menu

### LINKED ISSUE #2994: Feature Request: allow to pick the color of the tabs
# Description of the new feature/enhancement
Hello everybody, I think it is a good idea to give the user the possibility to pick colors of the currently opened tabs. The use case is the following: I use command line git and I work on multiple projects in parallel. This means that I have multiple tabs, each for a different git repo.

I want to be able to differentiate at a quick glance which tab is for which folder/repo/project/whatever and a tab color (if I can set one) is very strong visual cue. 

For example, Visual Studio Code has the [Peacock extension](https://marketplace.visualstudio.com/items?itemName=johnpapa.vscode-peacock), which gives you the possibility to color your editors differently, according to the project you are currently working on.

Visual Studio has similarly [SolutionColor](https://marketplace.visualstudio.com/items?itemName=Wumpf.SolutionColor)

# Proposed technical implementation details (optional)

What I imagine is that, if the user does a right click on a tab, a menu is shown, with an element called 'Choose a color' or something similar. When the element is clicked,  a simple color picker (ala OneNote) is displayed:

color_picker

Then there are two possibilities:
- the simple one is that the tabs are colored in the selected color and after the application is closed, this information is lost
- the more complex one: the application takes note of the current working dir of the shell, belonging to the tab and as long as the shell remains in this folder or a subfolder, the tab is colored in the chosen color. If this is even possible, one can store the folder/chosen color list in the settings and restore the tab colors, after the application is restarted.

If this is interesting to you, I would like to try to implement it. What are your thoughts?

Regards

### PR DESCRIPTION
**Pull request by gbaychev**
This commit introduces a context menu for Tab and a new item,
"Color...", which will display a color picker.

A flyout menu, containing a custom flyout, is attached to each tab. The
flyout displays a palette of 16 preset colors and includes a color
picker. When the user selects or clears color, an event is fired, which
is intercepted by the tab to which the flyout belongs.

The changing of the color is achieved by putting the selected color in
the resource dictionary of the tab, using well-defined dictionary keys
(e.g. TabViewItemHeaderBackground). Afterwards the visual state of the
tab is toggled, so that the color change is visible immediately.

Custom-colored tabs will be desaturated (somewhat) by alpha blending
them with the tab bar background.

The flyout menu also contains a 'Close' flyout item.

## Validation Steps Performed
I've validated the behavior manually: start the program via the start
menu. Right click on the tab -> Choose a tab color.

The color flyout is going to be shown. Click a color swatch or click
'Select a custom color' to use the color picker. Use the 'Clear the
current color' to remove the custom color.

Closes #2994. References #3327.

### CODE REVIEW

**Comment by mkitzan** [comment]
I was having code formatting problems too. @greg904 pointed out a tool for [code formatting](https://github.com/microsoft/terminal/pull/3762#issuecomment-559482434) in the repo.

**Comment by gbaychev** [comment]
@mkitzan Thanks, mate, yeah, I should've read the docs.

**Comment by DHowett-MSFT** [comment]
(Warning: we may be a little slow here. We're trying to get through our pull request backlog :smile: thanks for understanding!)

File: src/cascadia/TerminalApp/TabRowControl.xaml.orig [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                 AllowDropTabs="False">
+
+        <mux:TabView.TabStripFooter>
+            <mux:SplitButton
+                x:Name="NewTabButton"
+                Click="OnNewTabButtonClick"
+<<<<<<< Updated upstream
+=======
+                Background="Purple"
+>>>>>>> Stashed changes
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/TabRowControl.xaml.orig]: A conflict marker!

**Comment by gbaychev** [comment]
> 
> 
> (Warning: we may be a little slow here. We're trying to get through our pull request backlog 😄 thanks for understanding!)

No probs, thank you for getting back to me, most kind of you :)

File: src/cascadia/TerminalApp/TabRowControl.xaml.orig (hunk 1 above)
Comment (gbaychev) [reply src/cascadia/TerminalApp/TabRowControl.xaml.orig]: Damn, no idea how this file snuck by, it is removed now.

**Comment by zadjii-msft** [comment]
Okay so this is incredible. I'm having a [lot of fun](https://media2.giphy.com/media/PmABbbUe3IqUKSOIBV/giphy.gif?cid=790b7611a877b724d270a3b7e654b4ecdb2d54ef79bb2844&rid=giphy.gif) playing with this already.
image

The live updating with the color picker is also _awesome_
image

Looking forward to getting into the code :)

File: src/cascadia/TerminalApp/Tab.cpp [hunk 2]
```diff
@@ -335,4 +339,174 @@ void Tab::_AttachEventHandlersToPane(std::shared_ptr<Pane> pane)
...
+        _SetTabColor(newTabColor);
+    });
+
+    _tabColorPickup.ColorCleared([this]() {
+        _ResetTabColor();
+    });
+
+    chooseColorMenuItem.Click([this](auto&&, auto&&) {
+        _tabColorPickup.ShowAt(_tabViewItem);
+    });
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Tab.cpp]: (though we might need to do this work outside the ctor - if you pull, the `Tab::BindEventHandlers` method is called at the right time

File: src/cascadia/TerminalApp/ColorPickupFlyout.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+        auto okText = RS_(L"Ok");
+        auto customColorText = RS_(L"TabCustomColorChoose");
+        auto clearColorText = RS_(L"TabColorClear");
+
+        btnOk().Content(winrt::box_value(okText));
+        btnCustomColor().Content(winrt::box_value(customColorText));
+        btnClearColor().Content(winrt::box_value(clearColorText));
+
+        customColorPicker().ColorChanged([this](const auto&, const Windows::UI::Xaml::Controls::ColorChangedEventArgs& args) {
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/ColorPickupFlyout.cpp]: Like my earlier comment, we should probably be doing this like this comment suggests:
https://github.com/microsoft/terminal/issues/3919#issuecomment-564731215

```c++
customColorPicker().ColorChanged({ get_weak(), &ColorPickerFlyout::_ColorChangedHandler });
```

File: src/cascadia/TerminalApp/TabRowControl.xaml [hunk 4]
```diff
@@ -1,70 +1,70 @@
...
+                VerticalAlignment="Stretch"
+                HorizontalAlignment="Left"
+                Content="&#xE710;"
+                UseLayoutRounding="true"
+                FontFamily="Segoe MDL2 Assets"
+                FontWeight="SemiLight"
+                FontSize="12">
+                <!-- U+E710 is the fancy plus icon. -->
+                <mux:SplitButton.Resources>
+                    <!-- Override the SplitButton* resources to match the tab view's button's styles. -->
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TabRowControl.xaml]: amen 🙏

File: src/cascadia/TerminalApp/ColorFlyoutSubItem.idl [hunk 5]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+import "..\App.idl";
+
+//namespace TerminalApp
```
[the remark this answers is not in the document]
Comment (gbaychev) [reply src/cascadia/TerminalApp/ColorFlyoutSubItem.idl]: Yes, I'm going to delete this and a couple of other files that I've commited and forgotten to delete.

File: src/cascadia/TerminalApp/ColorPickupFlyout.xaml [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                    <Setter Property="MinHeight" Value="12"/>
+                </Style>
+            </Grid.Resources>
+            <ColorPicker x:Name="customColorPicker" 
+                     IsMoreButtonVisible="True"
+                     IsColorSliderVisible="False"
+                     IsColorChannelTextInputVisible="True"
+                     IsHexInputVisible="True"
+                     IsAlphaEnabled="False"
+                     IsAlphaSliderVisible="True"
```
[the remark this answers is not in the document]
Comment (gbaychev) [reply src/cascadia/TerminalApp/ColorPickupFlyout.xaml]: Absolutely

File: src/cascadia/TerminalApp/Tab.cpp [hunk 7]
```diff
@@ -2,8 +2,11 @@
 // Licensed under the MIT license.
 
 #include "pch.h"
+#include <LibraryResources.h>
+#include "ColorPickupFlyout.h"
 #include "Tab.h"
 #include "Utils.h"
+#include "winrt/Windows.UI.Popups.h"
```
[the remark this answers is not in the document]
Comment (gbaychev) [reply src/cascadia/TerminalApp/Tab.cpp]: Alright

File: src/cascadia/TerminalApp/Tab.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (gbaychev) [reply src/cascadia/TerminalApp/Tab.cpp]: I will change it. I'm not 100% what do you mean with your last sentence about `Tab::BindEventHandlers` but I'm going to check the code.

File: src/cascadia/TerminalApp/Tab.cpp [hunk 8]
```diff
@@ -335,4 +339,174 @@ void Tab::_AttachEventHandlersToPane(std::shared_ptr<Pane> pane)
...
+    auto clearAllSymbol = Controls::FontIcon{};
+    clearAllSymbol.FontFamily(Media::FontFamily{ L"Segoe MDL2 Assets" });
+    clearAllSymbol.Glyph(L"\xED62");
+
+    auto tabClose = RS_(L"TabClose");
+    closeTabMenuItem.Text(tabClose);
+    closeTabMenuItem.Icon(closeSymbol);
+
+    closeTabMenuItem.Click([this](auto&&, auto&&) {
+        ClosePane();
```
[the remark this answers is not in the document]
Comment (gbaychev) [reply src/cascadia/TerminalApp/Tab.cpp]: Yes, this closes the focused pane. I thought it was ok, if I can close a single pane. What do you mean precisely?

File: src/cascadia/TerminalApp/Tab.cpp [hunk 9]
```diff
@@ -335,4 +339,174 @@ void Tab::_AttachEventHandlersToPane(std::shared_ptr<Pane> pane)
...
+        deselectedTabColor.A = 64;
+        deselectedTabBrush.Color(deselectedTabColor);
+        _tabViewItem.Resources().Insert(winrt::box_value(L"TabViewItemHeaderBackgroundSelected"), selectedTabBrush);
+        _tabViewItem.Resources().Insert(winrt::box_value(L"TabViewItemHeaderBackground"), deselectedTabBrush);
+        _tabViewItem.Resources().Insert(winrt::box_value(L"TabViewItemHeaderBackgroundPointerOver"), selectedTabBrush);
+        _tabViewItem.Resources().Insert(winrt::box_value(L"TabViewItemHeaderForeground"), fontBrush);
+        _tabViewItem.Resources().Insert(winrt::box_value(L"TabViewItemHeaderForegroundSelected"), fontBrush);
+        _tabViewItem.Resources().Insert(winrt::box_value(L"TabViewItemHeaderForegroundPointerOver"), fontBrush);
+        //switch the visual state so that the colors are updated immediately
+        if (_focused)
```
[the remark this answers is not in the document]
Comment (gbaychev) [reply src/cascadia/TerminalApp/Tab.cpp]: Will do

File: src/cascadia/TerminalApp/Tab.cpp [hunk 10]
```diff
@@ -335,4 +339,174 @@ void Tab::_AttachEventHandlersToPane(std::shared_ptr<Pane> pane)
...
+            else
+            {
+                c[i] = std::pow((c[i] + 0.055f) / 1.055f, 2.4f);
+            }
+        }
+
+        // calculate the luminance of the current color and select a font
+        // color based on that
+        auto luminance = 0.2126f * c[0] + 0.7152f * c[1] + 0.0722f * c[2];
+        if (luminance > 0.179)
```
[the remark this answers is not in the document]
Comment (gbaychev) [reply src/cascadia/TerminalApp/Tab.cpp]: The coefficients are taken from [here](https://www.w3.org/TR/WCAG20/#relativeluminancedef) and I'm not sure that there are meaningful names.

File: src/cascadia/TerminalApp/ColorPickupFlyout.cpp (hunk 3 above)
Comment (gbaychev) [reply src/cascadia/TerminalApp/ColorPickupFlyout.cpp]: Roger

File: src/cascadia/TerminalApp/TabRowControl.xaml (hunk 4 above)
Comment (gbaychev) [reply src/cascadia/TerminalApp/TabRowControl.xaml]: Ah, I see, totally missed it.

**Comment by gbaychev** [comment]
Hello, guys, 
thank you very much for your feedback and I'm very happy that you've enjoyed the functionality that much! :)

I tried to fix the things noted above but this lead to Terminal crashing at startup, when the window title is set. I'm not sure what I've managed to screw up exactly, but currently I'm on a vacation and won't be able to check everything thoroughly this year. Just wanted to let you know.

Merry Christmas or whatever you are celebrating and enjoy the holidays.

**Comment by schorrm** [comment]
Could the profile JSON get hooked into this?
Add a TabColor (optional), and if a TabColor is setting is enabled color on create and on reset fire off a call to set?

**Comment by zadjii-msft** [comment]
@schorrm That's not a terrible idea, but I think it's honestly better suited for #3327. I've got a spec in progress for that issue which should be coming later this month, which _should_ address specifically that request.

<hr>

@gbaychev I think I know what's cause - I ran into something similar in #4032 (or one of my other branches before the new year). I think you need to change `TerminalPage::_GetActiveControl` to the following:
https://github.com/microsoft/terminal/blob/23f2cd22976beeabebd09e17f8bcd74fccf1a606/src/cascadia/TerminalApp/TerminalPage.cpp#L914-L919

If you want to push your updates, I can take a look myself and make sure.

**Comment by gbaychev** [comment]
Alright, most of the stuff is addressed. The weird crash was because I was trying to get a weak ptr in the constructor (what @zadjii-msft mentioned in a comment above). I made the `Tab::BindEventHandlers` private, created a method called `Tab::Initialize` and moved `Tab::_BindEventHandlers` and `Tab::_CreateContextMenu` into `Tab::Initialize`.

What is left is the updating of the status bar, will do it ASAP.

**Comment by dkter** [comment]
>I ran into something similar in #4032

@zadjii-msft I think you mentioned the wrong PR :)

File: src/cascadia/LocalTests_TerminalApp/pch.h.orig [hunk 11]
```diff
@@ -0,0 +1,63 @@
+/*++
```
Comment (zadjii-msft) [inline src/cascadia/LocalTests_TerminalApp/pch.h.orig]: Oh boy this file shouldn't be here

File: src/cascadia/TerminalApp/Tab.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Tab.cpp]: So we can probably leave this for a follow up, but I'd presume that "Close" means "Close Tab" and not "Close Pane". "Close Tab" would close all the panes in the tab, while "Close Pane" only closes the active one. Maybe we could have it just display "Close" if there's only one Pane, and display both "Close Tab" and "Close Pane" if there are multiple tabs?

I think this is fine for now, and something we can address in a follow up.

File: src/cascadia/TerminalApp/TabRowControl.xaml.orig [hunk 12]
```diff
@@ -0,0 +1,90 @@
+<<<<<<< HEAD
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TabRowControl.xaml.orig]: Oh no, this looks like another bad merge side-effect

File: src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj [hunk 13]
```diff
@@ -82,7 +88,7 @@
     <ClInclude Include="../WslDistroGenerator.h" />
     <ClInclude Include="../AzureCloudShellGenerator.h" />
     <ClInclude Include="../TelnetGenerator.h" />
-    <ClInclude Include="pch.h" />
+        <ClInclude Include="pch.h" />
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj]:
```suggestion
    <ClInclude Include="pch.h" />
```

File: src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj.orig [hunk 14]
```diff
@@ -0,0 +1,356 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj.orig]: Oh no, here's another

File: src/cascadia/TerminalApp/Tab.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Tab.cpp]: Oh you definitely got it right. Thanks!

**Comment by miniksa** [comment]
I did a merge here: https://github.com/microsoft/terminal/tree/dev/miniksa/tabs-color-fix

It looks like resources are already in the .resw and using the `RS_` macro.

**Comment by miniksa** [comment]
Hmmm, @DHowett-MSFT says I might be able to push a merge... trying.

**Comment by DHowett-MSFT** [comment]
I'm not extremely comfortable with taking non-client area changes as part of this; is it possible to pull it out as a followup? Sorry 🤕

**Comment by gbaychev** [comment]
Hello everybody, @miniksa @zadjii-msft ,
I think all the issues now are addressed and I've improved a lot of things. Now:
- there is a color helper library for color operations
- foreground color is better calculated
- the close button now closes the whole tab, not just the active pane
- tab split button styles are adjusted corresponding to the tab color
- the spurious *.orig files are deleted 
- I paint also the the title bar, so that the user can easily differentiate which tab is currently selected

This is how it looks:
titlebar


By the way, I'm sure you've noticed, but I wanted to let you know that it appears, that the master branch is currently broken. When the first tab is opened, a nullptr is read, when enumerating the tabs:

stack
msg

I'm looking forward to your feedback
Regards

**Comment by gbaychev** [comment]
@DHowett-MSFT sure, just a couple of questions
- should I still paint the status bar and the tab row control?
- do I need to remove the whole code or simply remove the firing of the events, which signals the painting of the title bar?
- what is the problem with painting the non-client area?

**Comment by zadjii-msft** [comment]
First: that looks awesome. I love it. 

However, I agree with @DHowett-MSFT, we should probably pull it from this PR for now. as cool as it is, I think we'd feel more comfortable with a more atomic PR for just the tab color picker & right-click menu, and add the titlebar colorization in a future PR.

I'd probably remove all the code for changing the color of the titlebar, as well as changing the background color of the tab row, MinMaxCloseControl, etc. I'd probably pull the events too for now. 

I'd probably leave the ColorHelper class, since that looks pretty good to me.

I'm working with @cinnamon-msft to draft a spec outlining our thoughts for more complete XAML theming of the whole window (including the titlebar) currently. It's only a draft for now, but a lot of my thoughts are over in #3327. In general, the draft already makes a lot of references to this PR, but I think when it comes to stylizing the whole titlebar, we'll want to make sure whatever functionality meshes well with the boatload of other requests we have. I'll certainly make sure to add a section on using the tab color picker to also set the color of the titlebar, because this is just such a slick feature.

**Comment by gbaychev** [comment]
@zadjii-msft thank you kindly for the explanation, sorry that I went overboard.
I ripped out the code, left the ColorHelper class. The only thing left for now is to wait for the master to get fixed, so that I can merge and see whether or not everything works properly.

P.S. Now I've noticed that I made mistake in my previous (and another before that) comment - I've written 'status bar', by which I actually meant the 'new tab split button' (🤦‍♂)  .  But since adapting the styles of the split button was part of your initial feedback, I'm going the leave the code as it is for now.
I'm curious whether you like the new behavior.

**Comment by gbaychev** [comment]
I did the requested changes, everything builds correctly and the tests are passing. The ball is in your field, guys.

File: src/cascadia/ut_app/TerminalApp.UnitTests.vcxproj [hunk 15]
```diff
@@ -44,56 +41,51 @@
...
   <ItemGroup>
     <Manifest Include="TerminalApp.Unit.Tests.manifest" />
   </ItemGroup>
-
   <!-- Careful reordering these. Some default props (contained in these files) are order sensitive. -->
   <Import Project="$(OpenConsoleDir)src\common.build.post.props" />
   <Import Project="$(OpenConsoleDir)src\common.build.tests.props" />
-
   <PropertyGroup>
-    <_CppWinrtBinRoot>&quot;$(OpenConsoleDir)$(Platform)\$(Configuration)\&quot;</_CppWinrtBinRoot>
```
[the remark this answers is not in the document]
Comment (gbaychev) [reply src/cascadia/ut_app/TerminalApp.UnitTests.vcxproj]: It appears that Visual Studio changed it behind my back. Should I revert that change?

[the remark this answers is not in the document]
Comment (gbaychev) [reply src/cascadia/ut_app/TerminalApp.UnitTests.vcxproj]: Eh, I reverted it ...

**Comment by DHowett-MSFT** [comment]
I am happy to fix these things myself to get this pull request merged. If you're okay with that, I can push to your branch :smile:

**Comment by DHowett-MSFT** [comment]
I meant for custom to say `Custom...`. There's a long-held tradition of menu items with `...` on them meaning "another thing will pop up when you click this" :smile:

**Comment by gbaychev** [comment]
Well, it appears that you pushed everything while I was asleep 😆 . Looks good. Very good point about the WUI namespace alias, I did not think at all about it. 

Did you by any chance check my comments in #4663? I've already mentioned that the coloring of the tab split button should go away and proposed how you can indicate which tab is selected.

**Comment by cinnamon-msft** [comment]
Hey @gbaychev, thanks so much for this awesome new feature! Unfortunately, we won't be able to merge this into v1 given that we declared feature completion last month. I would like to apologize for our delay in moving this out of the milestone, but we're excited to have this as our first feature in v1.1! :)

File: src/cascadia/TerminalApp/ColorHelper.cpp [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+}
+
+// Method Description:
+// Gets the readability of two colors according to
+// http://www.w3.org/TR/2008/REC-WCAG20-20081211/#contrast-ratiodef (WCAG Version 2)
+// Arguments:
+// - firstColor: the first color for the readability check (rgb)
+// - secondColor: the second color for the readability check (rgb)
+// Return Value:
+// - the readability of the colors according to  (WCAG Version 2)
```
Comment (jsoref) [inline src/cascadia/TerminalApp/ColorHelper.cpp]:
```suggestion
// - the readability of the colors according to (WCAG Version 2)
```

**Pull request merged**